### PR TITLE
Add dockerfile

### PR DESCRIPTION
**Pull request by smk762**
- Added dockerfile to use for explorer at https://github.com/smk762/nucleus-explorer/tree/v0.46.x
- Added `argparse` to `init.py` script to allow running non-interactive.
- Added `addrbook.json` to help bootstrap nodes
- Updated some modules to accomodate importing into https://github.com/smk762/callisto/tree/cosmos/v0.46.x (required for explorer compatibility).

Explorer is currently mostly functional, though it appears to be not sunning on the same chain as the existing nodes, even with the `addrbook.json` file. I think this might be by design - VCS stamping possibly enforces unmodified binaries for joining the network. If so, we probably need to set an update height using whatever mechanism is used here.

The explorer is otherwise functional.

image

If it is preferable to not update the version yet, I'll create a new branch of explorer without the nucleus container and deploy/attach on an existing nucleus server to serve in the meantime anyway.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (onur-ozkan) [inline Dockerfile:1]:
```suggestion
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+FROM golang:1.18-alpine AS builder
+RUN apk update && apk add --no-cache make git gcc libc-dev linux-headers python3-dev
+WORKDIR /nucleus
+COPY . /nucleus
+RUN rm -rf .git && git init
+RUN go mod tidy && go mod download
+RUN make clean & make build
```
Comment (onur-ozkan) [inline Dockerfile:8]:
```suggestion
RUN make clean build
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+RUN python3 ./init.py -b /nucleus/build/nucleusd -c /root/.nucleus --overwrite 
+COPY addrbook.json /root/.nucleus/config/addrbook.json
```
Comment (onur-ozkan) [inline Dockerfile:10]: Not every use case requires `init.py` to be run. We should provide a plain nucleus daemon and leave the rest to the users. How about simply copying `init.py` (without running) into the container?

Also `addrbook.json` file can be moved into https://github.com/KomodoPlatform/nucleus-assets repository.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+FROM alpine:latest
+WORKDIR /nucleus
+COPY --from=builder /nucleus/build/nucleusd /usr/bin/nucleusd
+COPY --from=builder /root/.nucleus /root/.nucleus
+# COPY --from=builder /nucleus/nucleus_conf.yaml /root/.nucleus/config.yaml
```
Comment (onur-ozkan) [inline Dockerfile:16]: We just need `nucleusd` and `init.py` from the builder.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
+
+FROM golang:1.18-alpine AS builder
+RUN apk update && apk add --no-cache make git gcc libc-dev linux-headers python3-dev
+WORKDIR /nucleus
+COPY . /nucleus
+RUN rm -rf .git && git init
```
Comment (onur-ozkan) [inline Dockerfile:6]: Why is this needed?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+RUN go mod tidy && go mod download
```
Comment (onur-ozkan) [inline Dockerfile:7]: > Updated some modules to accomodate importing into https://github.com/smk762/callisto/tree/cosmos/v0.46.x (required for explorer compatibility)

Can you add some more context about what exactly happens when we don't bump this dependencies?

File: Dockerfile (hunk 5 above)
Comment (smk762) [reply Dockerfile:5]: VCS stamping fails if there is no `.git` folder, this was a workaround. It is because the `.git` is a file, not a folder when its a submodule.

Comment (onur-ozkan) [reply Dockerfile:5]: Setting `GOFLAGS=-buildvcs=false` env should disable VCS.

File: Dockerfile (hunk 6 above)
Comment (onur-ozkan) [reply Dockerfile:6]: Thanks!

File: Dockerfile (hunk 3 above)
Comment (smk762) [reply Dockerfile:9]: copying init.py into container will mean needing to add some extra py related deps, and it wont work unless we also copy the app.toml template.

**Comment by smk762** [comment]
I've removed the dockerfile and the addrbook.json files. These are only needed for a specific use case and can be dealt with or accommodated as required elsewhere.

File: Dockerfile (hunk 5 above)
Comment (onur-ozkan) [reply Dockerfile:5]: > I think this is part of the "daemon version validation" consensus

I wonder if initializing fresh git context will cause any problem. Maybe there's a better solution instead of removing the actual git state?

**Comment by onur-ozkan** [comment]
> I've removed the dockerfile and the addrbook.json files. These are only needed for a specific use case and can be dealt with or accommodated as required elsewhere.

I am happy to merge this PR without them, did you push the changes (they still exists in the PR)?

**Review by onur-ozkan — APPROVED** [review]
